GDNative build failure on X11 armv7hl

**Godot version:**
Godot 3.0 (4a042b1)

**OS/device including version:**
Mageia 7, armv7hl

**Issue description:**
Attempting to build the current master on Mageia 7 (Cauldron)'s buildsystem, it passes fine on i586 and x86_64 but fails on armv7hl with the following error:
```
build_gdnative_api_struct(["modules/gdnative/include/gdnative_api_struct.gen.h", "modules/gdnative/gdnative_api_struct.gen.cpp"], ["modules/gdnative/gdnative_api.json"])
Compiling ==> modules/gdnative/gdnative.cpp
Compiling ==> modules/gdnative/register_types.cpp
Compiling ==> modules/gdnative/gdnative/pool_arrays.cpp
Compiling ==> modules/gdnative/gdnative/vector2.cpp
Compiling ==> modules/gdnative/gdnative/rid.cpp
Compiling ==> modules/gdnative/gdnative/variant.cpp
Compiling ==> modules/gdnative/gdnative/basis.cpp
Compiling ==> modules/gdnative/gdnative/quat.cpp
modules/gdnative/gdnative/variant.cpp: In function 'godot_variant godot_variant_call(godot_variant*, const godot_string*, const godot_variant**, godot_int, godot_variant_call_error*)':
modules/gdnative/gdnative/variant.cpp:40:81: internal compiler error: in assign_temp, at function.c:968
 #define memnew_placement_custom(m_placement, m_class, m_constr) _post_initialize(new (m_placement, sizeof(m_class), "") m_constr)
                                                                 ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
modules/gdnative/gdnative/variant.cpp:449:2: note: in expansion of macro 'memnew_placement_custom'
  memnew_placement_custom(dest, Variant, Variant(self->call(*method, args, p_argcount, error)));
  ^~~~~~~~~~~~~~~~~~~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://bugs.mageia.org/> for instructions.
scons: *** [modules/gdnative/gdnative/variant.x11.opt.tools.32.o] Error 1
scons: building terminated because of errors.
```

Build command:
```
/usr/bin/scons -j4 'CCFLAGS=-O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -march=armv7-a -mfloat-abi=hard -mfpu=vfpv3-d16' 'LINKFLAGS= -Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro -Wl,-O1 -Wl,--build-id -Wl,--enable-new-dtags' builtin_freetype=no builtin_libogg=no builtin_libpng=no builtin_libtheora=no builtin_libvorbis=no builtin_libvpx=no builtin_libwebp=no builtin_openssl=no builtin_opus=no builtin_pcre2=no builtin_zlib=no udev=yes progress=no p=x11 tools=yes target=release_debug
```

Full build log: https://hastebin.com/edufijugih
GCC version: 7.3.0

CC @karroffel

I just failed to build the latest version (older versions also don't work) on a Raspberry Pi 3b running Raspbian. 

Some system information:
```
$ uname -a
Linux raspberry 4.14.81-v7+ #1163 SMP Thu Nov 15 17:16:30 GMT 2018 armv7l GNU/Linux

$ gcc --version
gcc (Raspbian 6.3.0-18+rpi1+deb9u1) 6.3.0 20170516

$ glxinfo | grep "OpenGL"
...
OpenGL ES profile version string: OpenGL ES 3.0 Mesa 13.0.6
...
```


The build command I tried
```
scons p=x11 -j 3 tools=no target=release arch=arm64
```

I tried to use the workaround provided by @akien-mga by adding
```
env_gdnative.Append(CCFLAGS="-O3 -fno-tree-fre -fno-tree-dominator-opts -fno-tree-copy-prop -fno-tree-ccp -fno-code-hoisting -fno-tree-pre -fno-tree-vrp".split())  
```

to `modules/gdnative/SCsub` but I get the following error
```
[Initial build] Compiling ==> modules/gdnative/gdnative.cpp
g++: error: unrecognized command line option '-fno-code-hoisting'; did you mean '-Wno-type-hiding'?
[Initial build] Compiling ==> modules/gdnative/register_types.cpp
[Initial build] g++: error: unrecognized command line option '-fno-code-hoisting'; did you mean '-Wno-type-hiding'?
[Initial build] scons: *** [modules/gdnative/gdnative.x11.opt.arm64.o] Error 1
scons: *** [modules/gdnative/register_types.x11.opt.arm64.o] Error 1
Compiling ==> modules/gdnative/android/android_gdn.cpp
g++: error: unrecognized command line option '-fno-code-hoisting'; did you mean '-Wno-type-hiding'?
scons: *** [modules/gdnative/android/android_gdn.x11.opt.arm64.o] Error 1
scons: building terminated because of errors.

```

When I remove the flag `-fno-code-hoisting` I get lots of errors like this: 
```
In file included from thirdparty/libvpx/vp8/common/arm/neon/copymem_neon.c:11:0:
thirdparty/libvpx/vp8/common/arm/neon/copymem_neon.c: In function 'vp8_copy_mem8x4_neon':
/usr/lib/gcc/arm-linux-gnueabihf/6/include/arm_neon.h:9440:1: error: inlining failed in call to always_inline 'vst1_u8': target specific option mismatch
 vst1_u8 (uint8_t * __a, uint8x8_t __b)
thirdparty/libvpx/vp8/common/arm/neon/copymem_neon.c:23:9: note: called from here
         vst1_u8(dst, vtmp);
```

I am not really familiar with build systems and I already googled all these errors but nothing so far has worked. 
When searching for the flag `-fno-code-hoisting` it does not return any useful results at all. Are these godot-specific flags?

Any help would be appreciated.

--- a/modules/gdnative/gdnative/variant.cpp
+++ b/modules/gdnative/gdnative/variant.cpp
@@ -37,7 +37,21 @@
 extern "C" {
 #endif
 
+// Workaround GCC ICE on armv7hl which was affected GCC 6.0 up to 8.0 (GH-16100).
+// It was fixed upstream in 8.1, and a fix was backported to 7.4.
+// This can be removed once no supported distro ships with versions older than 7.4.
+#if defined(__arm__) && defined(__GNUC__) && !defined(__clang__) && \
+		(__GNUC__ == 6 || (__GNUC__ == 7 && __GNUC_MINOR__ < 4) || (__GNUC__ == 8 && __GNUC_MINOR__ < 1))
+#pragma GCC push_options
+#pragma GCC optimize("-O0")
+#endif
+
 #define memnew_placement_custom(m_placement, m_class, m_constr) _post_initialize(new (m_placement, sizeof(m_class), "") m_constr)
+
+#if defined(__arm__) && defined(__GNUC__) && !defined(__clang__) && \
+		(__GNUC__ == 6 || (__GNUC__ == 7 && __GNUC_MINOR__ < 4) || (__GNUC__ == 8 && __GNUC_MINOR__ < 1))
+#pragma GCC pop_options
+#endif
 
 // Constructors
 

